File: Stack/DynamicStack/DynamicStack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class DynamicStack:
    def __init__(self):
        self.top_node = None
        self.stack_size = 0

    def isEmpty(self):
        return self.top_node is None

    def push(self, value):
        new_node = Node(value)
        new_node.next = self.top_node
        self.top_node = new_node
        self.stack_size += 1

    def pop(self):
        if self.isEmpty():
            print("Stack Underflow! Cannot pop")
            return None
        popped_value = self.top_node.value
        self.top_node = self.top_node.next
        self.stack_size -= 1
        return popped_value

    def peek(self):
        if self.isEmpty():
            print("Stack is empty!")
            return None
        return self.top_node.value

    def size(self):
        return self.stack_size

    def print(self):
        if self.isEmpty():
            print("Stack is empty")
            return
        print("Stack (Top -> Bottom):")
        current = self.top_node
        while current:
            print(f"| {current.value} |")
            current = current.next
        print("-------")

if __name__ == "__main__":
    stack = DynamicStack()
    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.print()

    print("Popped:", stack.pop())
    print("Top element:", stack.peek())
    print("Current size:", stack.size())
    stack.print()

File: Stack/DynamicStack/DynamicStack.java
public class DynamicStack {
    private static class Node {
        int value;
        Node next;

        Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node top;
    private int size;

    public DynamicStack() {
        this.top = null;
        this.size = 0;
    }

    public boolean isEmpty() {
        return top == null;
    }

    public void push(int value) {
        Node newNode = new Node(value);
        newNode.next = top;
        top = newNode;
        size++;
    }

    public int pop() {
        if (isEmpty()) {
            System.out.println("Stack Underflow! Cannot pop");
            return -1;
        }
        int poppedValue = top.value;
        top = top.next;
        size--;
        return poppedValue;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Stack is empty!");
            return -1;
        }
        return top.value;
    }

    public int size() {
        return size;
    }

    public void print() {
        if (isEmpty()) {
            System.out.println("Stack is empty");
            return;
        }
        System.out.println("Stack (Top -> Bottom):");
        Node current = top;
        while (current != null) {
            System.out.println("| " + current.value + " |");
            current = current.next;
        }
        System.out.println("-------");
    }

    public static void main(String[] args) {
        DynamicStack stack = new DynamicStack();
        stack.push(10);
        stack.push(20);
        stack.push(30);
        stack.print();

        System.out.println("Popped: " + stack.pop());
        System.out.println("Top element: " + stack.peek());
        stack.print();
    }
}
